--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -1131,6 +_,7 @@
 
         // Folia start - region threading
         java.util.concurrent.CompletableFuture<Boolean> ret = new java.util.concurrent.CompletableFuture<>();
+        if (location.getWorld() == null) { ret.complete(false); return ret; } // Canvas - don't crash the server if a plugin provides a null world
         java.util.function.Consumer<Entity> run = (Entity nmsEntity) -> {
             boolean success = nmsEntity.teleportAsync(
                 ((CraftWorld)locationClone.getWorld()).getHandle(),
@@ -1238,7 +_,7 @@
 
     @Override
     public Set<org.bukkit.entity.Player> getTrackedPlayers() {
-        ChunkMap.TrackedEntity tracker = this.entity.moonrise$getTrackedEntity(); // Folia - region threading
+        ChunkMap.TrackedEntity tracker = this.getHandle().moonrise$getTrackedEntity(); // Folia - region threading // Canvas - region threading
         if (tracker == null) {
             return java.util.Collections.emptySet();
         }
