--- a/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -66,7 +_,7 @@
             // CraftBukkit end
             level.setBlockAndUpdate(pos, Blocks.DIRT.defaultBlockState());
         } else {
-            if (level.getMaxLocalRawBrightness(pos.above()) >= 9) {
+            if (io.canvasmc.canvas.Config.INSTANCE.disableGrassLightChecks || level.getMaxLocalRawBrightness(pos.above()) >= 9) { // Canvas - disable grass light checks config
                 BlockState blockState = this.defaultBlockState();
 
                 for (int i = 0; i < 4; i++) {
