--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -107,7 +_,7 @@
             } else {
                 double d = 3.0;
                 double d1 = 8.0;
-                double d2 = livingEntity.fallDistance;
+                double d2 = io.canvasmc.canvas.Config.INSTANCE.combat.mace.ignoreFallDistance ? 1.0D : Math.min(livingEntity.fallDistance, io.canvasmc.canvas.Config.INSTANCE.combat.mace.fallDistanceLimit == -1 ? livingEntity.fallDistance : io.canvasmc.canvas.Config.INSTANCE.combat.mace.fallDistanceLimit); // Canvas - mace config
                 double d3;
                 if (d2 <= 3.0) {
                     d3 = 4.0 * d2;
@@ -117,9 +_,10 @@
                     d3 = 22.0 + d2 - 8.0;
                 }
 
-                return livingEntity.level() instanceof ServerLevel serverLevel
+                return ((livingEntity.level() instanceof ServerLevel serverLevel
                     ? (float)(d3 + EnchantmentHelper.modifyFallBasedDamage(serverLevel, livingEntity.getWeaponItem(), target, damageSource, 0.0F) * d2)
-                    : (float)d3;
+                    : (float)d3)
+                    * (float) io.canvasmc.canvas.Config.INSTANCE.combat.mace.damageMultiplier);
             }
         } else {
             return 0.0F;
