--- a/net/minecraft/server/commands/SeedCommand.java
+++ b/net/minecraft/server/commands/SeedCommand.java
@@ -12,6 +_,16 @@
             long seed = commandContext.getSource().getLevel().getSeed();
             Component component = ComponentUtils.copyOnClickText(String.valueOf(seed));
             commandContext.getSource().sendSuccess(() -> Component.translatable("commands.seed.success", component), false);
+            // Leaf start - Matter - SecureSeed Command
+            if (io.canvasmc.canvas.Config.INSTANCE.secureSeed) {
+                su.plo.matter.Globals.setupGlobals(commandContext.getSource().getLevel());
+                String seedStr = su.plo.matter.Globals.seedToString(su.plo.matter.Globals.worldSeed);
+                Component featureSeedComponent = ComponentUtils.copyOnClickText(seedStr);
+
+                commandContext.getSource().sendSuccess(() -> Component.translatable(("Feature seed: %s"), featureSeedComponent), false);
+            }
+            // Leaf end - Matter - SecureSeed Command
+            
             return (int)seed;
         }));
     }
