--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2872,4 +_,14 @@
      */
     void clearBlockHighlights();
     // Purpur end - Debug Marker API
+    // Canvas start - expand region api
+
+    /**
+     * Returns the region on the current thraed, or null if we are on the global tick thread
+     * <br>
+     * This will throw if not on a tick runner
+     * @return the current region
+     */
+    @Nullable io.canvasmc.canvas.threadedregions.WorldRegionData getCurrentRegion();
+    // Canvas end
 }
