--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -15,6 +_,7 @@
     private static final long XZ_RANGE = 32L;
     private static final long Y_RANGE = 16L;
     public static final int MAX_DISTANCE_TO_WANTED_ITEM = 32;
+    private static final double MAX_DIST_SQ = (double) MAX_DISTANCE_TO_WANTED_ITEM * MAX_DISTANCE_TO_WANTED_ITEM; // Leaf - quick sort
 
     @Override
     public Set<MemoryModuleType<?>> requires() {
@@ -24,8 +_,16 @@
     @Override
     protected void doTick(ServerLevel level, Mob entity) {
         Brain<?> brain = entity.getBrain();
-        List<ItemEntity> entitiesOfClass = level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities
-        entitiesOfClass.sort(Comparator.comparingDouble(entity::distanceToSqr));
+        // Leaf start - fast bit radix sort
+        net.minecraft.core.Position pos = entity.position();
+        double x = pos.x();
+        double y = pos.y();
+        double z = pos.z();
+        net.minecraft.world.phys.AABB boundingBox = entity.getBoundingBox().inflate(32.0, 16.0, 32.0);
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity> entitiesOfClass = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel) level).moonrise$getEntityLookup().getEntities(ItemEntity.class, null, boundingBox, entitiesOfClass, (ItemEntity itemEntity) -> itemEntity.distanceToSqr(x, y, z) < MAX_DIST_SQ && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities
+        level.fastBitRadixSort.sort(entitiesOfClass.elements(), entitiesOfClass.size(), pos);
+        // Leaf end - fast bit radix sort
         // Paper start - Perf: remove streams from hot code
         ItemEntity nearest = null;
         for (final ItemEntity itemEntity : entitiesOfClass) {
